### PR TITLE
[go] Add Go version selection via `go.mod`

### PR DESCRIPTION
**Pull request by styfle**
#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [x] This PR has a concise title and thorough description useful to a reviewer
- [x] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/now-go/go-helpers.ts [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
+const archMap = new Map([
+  ['x64', 'amd64'],
+  ['x86', '386'],
+]);
 const platformMap = new Map([['win32', 'windows']]);
-
-// Location where the `go` binary will be installed after `postinstall`
-const GO_DIR = join(__dirname, 'go');
-const GO_BIN = join(GO_DIR, 'bin', 'go');
+export const cachDir = '.vercel-go-cache';
```
Comment (TooTallNate) [inline packages/now-go/go-helpers.ts]: Typo (needs to be updated in all places)?

```suggestion
export const cacheDir = '.vercel-go-cache';
```

File: packages/now-go/go-helpers.ts [hunk 2]
```diff
@@ -156,5 +166,20 @@ export async function downloadGo(
...
     });
   }
-  return createGo(dir, platform, arch);
+  return createGo(workPath, dir, platform, arch);
+}
+
+async function parseGoVersion(modulePath: string): Promise<string> {
+  let version = '1.16';
+  const file = join(modulePath, 'go.mod');
+  if (!existsSync(file)) {
```
Comment (TooTallNate) [inline packages/now-go/go-helpers.ts]: I'd prefer to `try/catch` the `readFile()` function and print the `debug()` line when `err.code === 'ENOENT'`.

File: packages/now-go/go-helpers.ts [hunk 3]
```diff
@@ -156,5 +166,20 @@ export async function downloadGo(
...
+    debug(`go.mod not found: ${file}`);
+    return version;
+  }
+  const content = await readFile(file, 'utf8');
+  const match = /^go (\d+\.\d+)\.?$/gm.exec(content);
+  if (match && match[1]) {
```
Comment (TooTallNate) [inline packages/now-go/go-helpers.ts]:
```suggestion
  if (match?[1]) {
```

File: packages/now-go/index.ts [hunk 4]
```diff
@@ -136,8 +142,12 @@ Learn more: https://github.com/golang/go/wiki/Modules
     if (fileName in downloadedFiles) {
       goModAbsPathDir = dirname(downloadedFiles[fileName].fsPath);
       debug(`Found ${fileName} file in "${goModAbsPathDir}"`);
+    } else if ('api/go.mod' in downloadedFiles) {
```
Comment (TooTallNate) [inline packages/now-go/index.ts:145]: What if I have `api/foo/go.mod`? I feel like we should be using `scanParentDirs()` or whatever that function is called to iterate up and find the `go.mod` file.

Comment (styfle) [reply packages/now-go/index.ts:145]: @TooTallNate We don't allow that behavior and I don't think Go does either. This was really just a feature for users to separate backend and frontend (backend being everything in the api directory).

**Review by TooTallNate — DISMISSED** [review]

File: packages/now-go/go-helpers.ts [hunk 5]
```diff
@@ -156,5 +163,26 @@ export async function downloadGo(
...
+    if (major === 1 && minor > 12) {
+      return `${major}.${minor}`;
+    } else {
+      debug(`Failed to parse version or its too old: ${file}`);
+    }
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      debug(`File not found: ${file}`);
+    }
+    throw err;
```
Comment (TooTallNate) [inline packages/now-go/go-helpers.ts]:
```suggestion
    } else {
      throw err;
    }
```

**Review by TooTallNate — CHANGES_REQUESTED** [review]

**Review by TooTallNate — APPROVED** [review]

**Pull request merged**